File: backend/ai_interview_bot/services/code_executor.py
import os
import requests
from dotenv import load_dotenv

load_dotenv()

def execute_code(code, language="python3", versionIndex="4"):
    client_id = os.getenv("JDOODLE_CLIENT_ID")
    client_secret = os.getenv("JDOODLE_CLIENT_SECRET")
    
    if not client_id or not client_secret:
        return {"error": "JDoodle credentials not configured"}
    
    payload = {
        "clientId": client_id,
        "clientSecret": client_secret,
        "script": code,
        "language": language,
        "versionIndex": versionIndex
    }
    
    try:
        response = requests.post("https://api.jdoodle.com/v1/execute", json=payload)
        response.raise_for_status()
        data = response.json()
        
        if data.get("statusCode") == 200:
            return {
                "output": data.get("output", ""),
                "memory": data.get("memory", ""),
                "cpuTime": data.get("cpuTime", "")
            }
        else:
            return {"error": f"JDoodle execution failed: {data.get('statusCode', 'Unknown error')}"}
    except requests.RequestException as e:
        return {"error": f"Request to JDoodle failed: {str(e)}"}
    except Exception as e:
        return {"error": f"Unexpected error: {str(e)}"}

File: frontend/services/codeExecutor.ts
// Code Executor Service - Integrates with JDoodle API (via backend)

const BACKEND_API = process.env.NEXT_PUBLIC_API_URL || "http://127.0.0.1:8000";

export interface ExecutionResult {
  output: string;
  error?: string;
  cpuTime?: string;
  memory?: string;
  status: "success" | "error" | "timeout";
}

export interface CodeSubmission {
  code: string;
  language: string;
  sessionId: string;
}

export const executeCode = async (
  code: string,
  language: string,
  sessionId: string
): Promise<ExecutionResult> => {
  try {
    const langInfo = SUPPORTED_LANGUAGES.find(l => l.id === language);
    
    const payload = {
      code,
      language,
      sessionId,
      versionIndex: "3",
    };
    
    
    const response = await fetch(`${BACKEND_API}/api/execute-code`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify(payload),
    });

    if (!response.ok) {
      const errorData = await response.text();
      console.error(`Backend error ${response.status}:`, errorData);
      
      return {
        output: "",
        error: `Execution failed: ${response.status} - ${errorData}`,
        status: "error",
      };
    }

    const data = await response.json();

    return {
      output: data.output || "",
      error: data.error || undefined,
      cpuTime: data.cpuTime,
      memory: data.memory,
      status: data.error ? "error" : "success",
    };
  } catch (error) {
    console.error("Code execution request error:", error);
    return {
      output: "",
      error: `Execution error: ${error instanceof Error ? error.message : "Unknown error"}`,
      status: "error",
    };
  }
};

export const SUPPORTED_LANGUAGES = [
  { id: "javascript", name: "JavaScript", version: 18 },
  { id: "python3", name: "Python 3", version: 3 },
  { id: "java", name: "Java", version: 14 },
  { id: "cpp14", name: "C++", version: 14 },
  { id: "csharp", name: "C#", version: 8 },
];

export const getLanguageId = (name: string): string => {
  const lang = SUPPORTED_LANGUAGES.find(
    (l) => l.name.toLowerCase() === name.toLowerCase()
  );
  return lang?.id || "javascript";
};
